wiki/osu!_File_Formats/Osu_(file_format)/en.md

```diff
@@ -176,90 +179,206 @@ Example of an inherited Timing Point:
 
-*addition (sampleSet:additions:customIndex:sampleVolume:filename)* is optional and defines the samplesets of the hitobject. It defaults to "0:0:0:0:". *sampleSet (Integer)* changes the sampleset of the object, and *addition (Integer)* changes the sampleset for additions (whistle, finish, clap). The values for these are:
+*filename* (String) names an audio file in the folder to play instead of sounds from sample sets, relative to the beatmap's directory.
 
-0 = Auto, 1 = Normal, 2 = Soft, 3 = Drum
+Unconfirmed: A comma in the *filename* may break everything.
 
-*customIndex (Integer)* is the custom sampleset index, e.g. 3 for `normal-3.wav`. *sampleVolume (Integer)* is the volume of the sample, 0-100 (percent). *filename (String)* names an audio file in the folder to play instead of sounds from samplesets.
+### Hit Circles
 
-**Slider Syntax:**
+A hit circle is a single hit in all osu! game modes.
+
+**Syntax**: `x,y,time,type,hitSound,addition`
+
+**Example**: `164,260,2434,1,0,0:0:0:0:`
+
+### Sliders
 
 A slider also creates droplets in Catch the Beat, yellow drumrolls in Taiko, and does not appear in osu!mania.
 
-`x,y,time,type,hitSound,sliderType|curvePoints,repeat,pixelLength,edgeHitsounds,edgeAdditions,addition`
+**Syntax**: `x,y,time,type,hitSound,sliderType|curvePoints,repeat,pixelLength,edgeHitsounds,edgeAdditions,addition`
+
+**Example**: `424,96,66,2,0,B|380:120|332:96|332:96|304:124,1,130,2|0,0:0|0:0,0:0:0:0:`
+
+#### Path
+
+*sliderType* will be `L` (linear), `P` (perfect), `B` (Bezier), or `C` (Catmull, deprecated).
+
+*curvePoints (x:y|...)* is a series of `|`-separated coordinates describing the control points of the slider.
+
+A **linear** slider has a start, specified in *x* and *y* from the common fields, and an end point specified in *curvePoints*. For example `L|100:200`.
+
+A **perfect** circle slider is defined by three points. In that order: start, pass-through, and end. *x* and *y* define the start point, and *curvePoints* defines the pass-through and end point. For example `P|250:200|200:15`.
+
+A perfect circle slider could be represented as a center point, a radius, and two angles for convenience. See this source code for the conversion algorithm: [CircularArcApproximator.cs](https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/CircularArcApproximator.cs).
+
+A **Bézier** slider is made of one or many Bézier curves, sharing common ends. The degree of each curve is arbitrary. The first control points is defined with *x* and *y*, and the other ones by *curvePoints*.
+
+To identify the separation between two curves, the intersection point is repeated. Consider the sequence ABCDDEFFG. You would get the 3 Bézier curves: ABCD (cubic), DEF (quadratic) , FG (linear).
 
-`424,96,66,2,0,B|380:120|332:96|332:96|304:124,1,130,2|0,0:0|0:0,0:0:0:0:`
+Note that this representation is ambiguous, because ABBBC could actually represent ABBBC, AB+BBC, AB+B+BC, or AB+BBC.
```

Comment:
not sure if this note is important; you'd never see three in a row anyway

Reply:
This is an edge case, and I did find a beatmap with ABB. If we were consistent, it's would become AB+B, which is the same as AB. However, it might as well be a quadratic curve ABB.

My point was that this edge case means the editor must ensure you won't see 2 identical control points in a row for a given curve. The cubic curve ABBC is perfectly reasonable, so I'll focus on it rather than use ABBBC which is more convoluted. What do you think?

Reply:
oh, I just meant the point about three in a row. beatmaps will never use more than two duplicate points for the reasons you described

I think you can just remove this line and the rest of the explanation is fine. It's true that this implementation does not do justice to all possible bezier curves, but it's not worth mentioning here because it's beyond the scope of osu! (and mapping with sliders). If the reader already knows what bezier curves are, they will understand it anyway, just as you have